front/components/assistant/conversation/input_bar/editor/useMentionDropdown.tsx

```diff
@@ -0,0 +1,278 @@
+import type { Editor } from "@tiptap/react";
+import { useCallback, useEffect, useRef, useState } from "react";
+
+import type {
+  EditorSuggestion,
+  EditorSuggestions,
+} from "@app/components/assistant/conversation/input_bar/editor/suggestion";
+import { filterSuggestions } from "@app/components/assistant/conversation/input_bar/editor/suggestion";
+
+interface CommandFunction {
+  (props: { id: string; label: string }): void;
+}
+
+interface Range {
+  from: number;
+  to: number;
+}
+
+interface SuggestionProps {
+  command: CommandFunction;
+  range: Range;
+  query: string;
+  clientRect: () => DOMRect | null;
+}
+
+interface MentionDropdownState {
+  isOpen: boolean;
+  query: string;
+  suggestions: EditorSuggestion[];
+  selectedIndex: number;
+  triggerRect: DOMRect | null;
+  isLoading: boolean;
+}
+
+export const useMentionDropdown = (
+  editorSuggestions: EditorSuggestions,
+  editorRef: React.MutableRefObject<Editor | null>
+) => {
+  const [state, setState] = useState<MentionDropdownState>({
```

Comment:
Any reason why we want to have one big state object? 🤔

Reply:
Seen IRL, matter of readability (to me this is more readable, even though not everyone would agree on this..)